src/substructureproperties.cxx

```diff
@@ -407,11 +407,11 @@ void GetCMProp(Options &opt, const Int_t nbodies, Particle *Part, Int_t ngroup,
     Int_t ii,icmv;
     Int_t RV_num;
     Double_t virval=log(opt.virlevel*opt.rhobg);
-    Double_t m200val=log(opt.rhobg/opt.Omega_m*200.0);
+    Double_t m200val=log(opt.rhocrit*200.0);
     Double_t m200mval=log(opt.rhobg*200.0);
-    Double_t mBN98val=log(opt.virBN98*opt.rhobg);
+    Double_t mBN98val=log(opt.virBN98*opt.rhocrit);
```

Comment:
Is this intended? I don't know the physical meaning behind rhocrit and rhobg, but they are clearly different values which seemed to have been computed differently in different parts of the code. I see that this pull is clearly trying to fix that, but this particular change (and that in line 2216 below) looks more like yet a different problem being solved, so I just wanted to make sure it was effectively intended.

Reply:
Thanks for the thorough review! I agree with moving the duplicated code into a separate function. I guess we'll have to discuss how far we're prepared to diverge from Pascal's repository.

rhobg is the mean mass density of the universe and rhocrit is the critical density above which the universe would collapse in on itself. There is also a density parameter omega which is defined as omega=rhobg/rhocrit. All three quantities vary with time. The variable omega_m in the code is the density parameter computed at the present day (redshift z=0 or expansion factor a=1) but the code was using it to compute rhocrit from rhobg at earlier times, which was wrong. The changes at lines 410 and 414 are part of the fix for this - instead of computing rhocrit inline when we need it, we compute it once and store it.

The change to mBN98val does look like a separate bug fix. I believe the new formula is correct (and the output agrees with the EAGLE code in my tests) but I'll check the BN98 paper to make sure.

Reply:
The quantity opt.virBN98 is delta_c in formula 6 in [Bryan and Norman (1998)](http://adsabs.harvard.edu/abs/1998ApJ...495...80B). The text states that it should be multiplied by the critical density to get the mean density within the virial radius, so the changes to mBN98val are fixing a bug.